File: ruby/palindrome-products/palindrome_products.rb
# Palindromes class comment
class Palindromes
  attr_reader :largest, :smallest
  def initialize(max_factor:, min_factor: 1)
    @max = max_factor
    @min = min_factor
  end

  def generate
    pal = Hash.new { |h, k| h[k] = [] }
    [*@min..@max].repeated_combination(2) do |m, n|
      num = m * n
      pal[num].push([m, n]) if num.to_s == num.to_s.reverse
    end

    min_max = Struct.new(:value, :factors)

    large_value, large_factors = pal.max
    @largest = min_max.new(large_value, large_factors)

    small_value, small_factors = pal.min
    @smallest = min_max.new(small_value, small_factors)
  end
end

File: python/palindrome-products/palindrome_products.py
def largest_palindrome(max_factor, min_factor):
    if min_factor > max_factor:
        raise ValueError("invalid factor")
    else:
        palindromes = get_palindromes(min_factor, max_factor + 1)
        if bool(len(palindromes)):
            value = max(palindromes)
            factors = get_factors(min_factor, max_factor, value)
        else:
            value, factors = empty_response()

        return value, factors


def smallest_palindrome(max_factor, min_factor):
    if min_factor > max_factor:
        raise ValueError("invalid factor")
    else:
        palindromes = get_palindromes(min_factor, max_factor + 1)
        if bool(len(palindromes)):
            value = min(palindromes)
            factors = get_factors(min_factor, max_factor, value)
        else:
            value, factors = empty_response()

        return value, factors


def get_factors(min_factor, max_factor, value):
    factors = set()
    for a in range(min_factor, max_factor + 1):
        for b in range(a, max_factor + 1):
            if a * b == value:
                factors.add((a, b))
    return factors


def get_palindromes(start, end):
    palindromes = []
    for i in range(start, end):
        for j in range(i, end):
            number = i * j
            # only add palindromes number
            if str(number) == str(number)[::-1]:
                palindromes.append(number)
    return palindromes


def empty_response():
    return None, []


def version():
    VERSION = 1
